server/api/events.go

```diff
@@ -0,0 +1,259 @@
+	Location    string `json:"location,omitempty"`
+	ChannelID   string `json:"channel_id"`
+}
+
+func (cep createEventPayload) ToRemoteEvent(loc *time.Location) (*remote.Event, error) {
+	var evt remote.Event
+
+	evt.IsAllDay = cep.AllDay
+
+	start, err := cep.parseStartTime(loc)
+	if err != nil {
+		return nil, errors.Wrap(err, "error parsing start time")
+	}
+
+	end, err := cep.parseEndTime(loc)
+	if err != nil {
+		return nil, errors.Wrap(err, "error parsing start time")
+	}
+
+	if !cep.AllDay {
+		evt.Start = &remote.DateTime{
+			DateTime: start.Format(remote.RFC3339NanoNoTimezone),
+			TimeZone: loc.String(),
+		}
+		evt.End = &remote.DateTime{
+			DateTime: end.Format(remote.RFC3339NanoNoTimezone),
+			TimeZone: loc.String(),
+		}
+	} else {
+		date, err := cep.parseDate(loc)
+		if err != nil {
+			return nil, errors.Wrap(err, "error parsing date")
+		}
+
+		evt.Start = &remote.DateTime{
+			DateTime: time.Date(date.Year(), date.Month(), date.Day(), 0, 0, 0, 0, loc).Format(remote.RFC3339NanoNoTimezone),
+			TimeZone: loc.String(),
+		}
+		evt.End = &remote.DateTime{
+			DateTime: time.Date(date.Year(), date.Month(), date.Day(), 23, 59, 59, 99, loc).Format(remote.RFC3339NanoNoTimezone),
+			TimeZone: loc.String(),
+		}
+	}
+
+	if cep.Description != "" {
+		evt.Body = &remote.ItemBody{
+			Content:     cep.Description,
+			ContentType: "text/plain",
+		}
+	}
+	evt.Subject = cep.Subject
+	if cep.Location != "" {
+		evt.Location = &remote.Location{
+			DisplayName: cep.Location,
+		}
+	}
+
+	return &evt, nil
+}
+
+func (cep createEventPayload) parseStartTime(loc *time.Location) (time.Time, error) {
+	return time.ParseInLocation(createEventDateTimeFormat, fmt.Sprintf("%s %s", cep.Date, cep.StartTime), loc)
+}
+
+func (cep createEventPayload) parseEndTime(loc *time.Location) (time.Time, error) {
+	return time.ParseInLocation(createEventDateTimeFormat, fmt.Sprintf("%s %s", cep.Date, cep.EndTime), loc)
+}
+
+func (cep createEventPayload) parseDate(loc *time.Location) (time.Time, error) {
+	return time.ParseInLocation(createEventDateFormat, cep.Date, loc)
+}
+
+func (cep createEventPayload) IsValid(loc *time.Location) error {
+	if cep.Subject == "" {
+		return fmt.Errorf("subject must not be empty")
+	}
+
+	if cep.Date == "" {
+		return fmt.Errorf("date must not be empty")
+	}
+
+	_, err := cep.parseDate(loc)
+	if err != nil {
+		return fmt.Errorf("invalid date")
+	}
+
+	if cep.StartTime == "" && cep.EndTime == "" && !cep.AllDay {
+		return fmt.Errorf("start time/end time must be set or event should last all day")
+	}
+
+	start, err := cep.parseStartTime(loc)
+	if err != nil {
+		return fmt.Errorf("please use a valid start time")
+	}
+
+	end, err := cep.parseEndTime(loc)
+	if err != nil {
+		return fmt.Errorf("please use a valid end time")
+	}
+
+	if start.After(end) {
+		return fmt.Errorf("end date cannot be earlier than start date")
+	}
+
+	return nil
+}
+
+func (api *api) createEvent(w http.ResponseWriter, r *http.Request) {
```

Comment:
I wonder if we can get this function under test. Also note that I haven't thoroughly reviewed this file yet. In general LGTM